### ISSUE TITLE
Intersection with weak-typed property in common doesn't allow assignment from expected sources

### ISSUE THREAD

**Issue by mheavenor** [microsoft/TypeScript#32557]
I originally ran into this issue because I was trying to extend Material UI interfaces - specifically the classes property. Upon investigation I found the following behavior seemingly inconsistent.

In the example below, if `valid2` was typed as either `A` or `B` rather than `C` it would be considered invalid because `foo` has an unknown property.

So if `valid2` is properly valid as per the Typescript spec and the types of `a` and `b` are both optional in `foo`, why are `invalid1` and `invalid2` invalid?

```ts
interface A {
    foo?: {
        a?: string;
    }
}

interface B {
    foo?: {
        b?: string
    }
}

type C = A & B; // Type C has a foo property that has properties a and b

const invalid1: C = {
    foo: {
        a: ""
    }
};

const invalid2: C = {
    foo: {
        b: ""
    }
};

const valid1: C = {
    foo: {}
};

const valid2: C = {
    foo: {
      a: "",
      b: ""
    }
};
```

[Playground](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgILIN4FgBQz-IwD2RA-AFya4E3JwXIDOYUoA5gNzUEC+ufOXKEixEKAEJU8BYmUrZptZACMGzViDbd8AgbjABPAA4oAwsgC8aZADJk4roJwIiIZslAA3OABtgAEwBGSnMrBRpZeW1aOEoAIjjo3UdcFzcwDxBvP38AJhDLKQiSKMVaZXjEsuTcVNd3bIDg5FCimRLMGqc0ht8A-JbC8PaiUqVY5ASAGmj8CsmqmmSgA)

**Comment by jcalz** [comment]
This is [weak type detection](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#weak-type-detection) as described in #16047.  There's this:

>Intersection types are handled specially as weak type targets. Individual constituents of an intersection are exempt from the weak type check. However, if all constituents of an intersection are weak, then the weak type check uses the combined properties of the intersection as the target.

which would imply that everything should be fine with your examples... but I guess the nested property thing is the issue; those aren't aggressively/recursively intersected I guess?

```ts
type X = {a?: string} & {b?: string};
const x: X = {b: "okay"}; // okay

type Y = {foo: X};
const y: Y = {foo: {b: "okay"}}; //okay

type Z = {foo: {a?: string}} & {foo: {b?: string}};
const z: Z = {foo: {b: "okay"}}; // error!
```

[Playground](http://www.typescriptlang.org/play/#code/C4TwDgpgBAGlC8UDeBDA-ALigZ2AJwEsA7AcwF8oAyZAI0x32PIG4BYAKAGMB7I3KAB5Y4iJDSwAibgGsUICWWZQA9Mqgy5HDqEhQAmgmQAzbt2GKOPPsCggsB0SbO1JG+WUUrlbrex3QALUMkJyxUelxCUg8qY1MwuixIpg82Ll5+AC8sIMd4lygpWXdPVSgIPDxuPABCXw4gA)

🤷

**Comment by RyanCavanaugh** [comment]
This was fixed in 3.7
